### DISCUSSION TITLE
The CHANGES button

### DISCUSSION THREAD

**Discussion by bkw777** [bkw777/mainline#224] [General]
I figured I'd explain a little about removing the changelog button. (I was rather short with my reply to a recent issue where I just said "no" and closed the issue. Sorry guy, I was real tired right at that moment and not well-behaved.)

1 - Recently another developer had the idea to "lazy load" the changelog files, meaning don't download them up front along with the index.html's but only download one if and when a user actually clicks on one. That was a good idea. On startup the app already has to do 100-ish or more http transactions at minimum just to update all the index.html's, and doubling that with the changelog files really slows everything down, and for no justifiable reason because 99% of the files don't get viewed every time they are re-downloaded, even IF you do look at every single one at least once.

Well what is the difference between "lazy downloading" the file on-demand and then launching your text editor to view it, and launching your browser which does the same thing, downloads it on the spot and displays it? Pretty much just that you could edit your local cached copy of the changes file vs just view it in a browser.

2 - We need the PPA button *anyway* totally seperately from the changelog file, because there is other info that goes along with each build besides just the changelog. It also helps with the bug reports about missing kernels that are really just failed builds. You can hit the ppa button as a quick way to check the source site to check any time you think a version might be erroneously missing. (I'm also planning on adding a toggle to display all the failed builds in the ui but greyed out or something. But until that's working, at least having a built-in way to jump right to the source just makes it more convenient to use.) Basically the whole function of the app is totally dependant on the behavior and function of the mainline-ppa site which I don't control or even have any contacts with or anything. Any user of this app really always needs to be able to jump to the ppa site frequently and frictionlessly to double-check is some kernel even up there at all yet, is it a failed build, is the site even up, did the site do something weird that breaks the apps ability to scrape it, etc... Since, at least if you buy my argument, the PPA button is needed regardless, then the changes button becomes just a redundant subset. I don't want two buttons where on is just a redundant subset of the other.

3 - The app needs the internet to do it's core job of downloading the .deb files, so I see no off-line operation argument. If you can install one of the listed kernels, then you can view the changelog via the PPA button. It means you can't see the changelogs while off-line. I can not concoct a scenario no matter how tortured where it is some important function to have access to all the changelog files off-line. If you are a kernel developer or something, then you can have that all kinds of other ways like with a clone of a git repo. There's no way that matters to anyone else. Or, go ahead and explain the use-case. What am I missing? The single extra button click to click on the changes file after clicking on the PPA button? The wins from losing the downloads and getting the entire directory of info vs one file outweigh that.

So that's my thinking on that. I guess some people used it and liked it, but I just can not see how it's worth keeping. I don't see how it's materially better than launching a browser at the parent web page. It's different but so ever so slightly that I just don't see how it matters, but having streamlined access to the rest of the build info and the rest of the site does matter, and having both is just blech.

I used this to try out the discussion feature because I want whoever has something to say about it to have a good way to do it, and not just in an issue whiuch I just close WONTFIX immediately or private email instead of a public discussion. So go ahead and discuss! Maybe I'll be less of a dick here than in an issue.

**Comment by SamuelMarks** [comment]
Here's my thinking, if you hit https://www.kernel.org/feeds/kdist.xml then you'll know the CHANGELOG link for each kernel version that has a CHANGELOG link and be able to download it on-demand when the user clicks the CHANGELOG button. Or since that list seems kinda small, hitting each of these:
- https://mirrors.edge.kernel.org/pub/linux/kernel/v4.x/
- https://mirrors.edge.kernel.org/pub/linux/kernel/v5.x/
- https://mirrors.edge.kernel.org/pub/linux/kernel/v6.x/

Will give you all the information needed to know: which version is available and where to download it from and whether it has a CHANGELOG associated. 3 HTTP requests for the last 8 years of Linux kernel versions.

Unless I'm missing something?

I.e., lazy loading as the 1st suggestion you referenced seems good.
PS: To answer the why question, it also saves going to the website and checking out the changelog when deciding whether or not to install that kernel version. One click show me the changes seems like a nice GUI feature.

**Comment by bkw777** [comment]
I don't understand. The PPA button already takes you right to the same file that the old CHANGES button did.

Highlight any kernel, then hit the PPA button. One of the links on the page that pops up is CHANGES. That is the file that used to be downloaded and cached and displayed by the old changes button.

General kernel changes info could come from anywhere, but info about availability can only come from the ppa site. Lots of kernels are published upstream that are never built or published by the ubuntu mainline ppa.

So at the moment I'm only talking about the change from the old changes button to the current ppa button. 

And I assert that that change is practically none. You press a button, you get the changelog. Yes it is one extra click vs previously, but it saves many other clicks elsewhere because otherwise you have to go outside the app and google the mainline ppa just to check why you don't see 6.0.2 or whatever.

**Comment by bkw777** [comment]
If you have not yet selected any kernel, the the ppa button goes to the index of all kernels.

If you have highlighted a kernel, it takes you to that kernel's page, but you still have ready acces to the parent index because every kernel page has a parent link.

**Comment by bkw777** [comment]
Maybe the button should just be labelled "Status/Changelog" instead of "PPA" ?

Or with those labels, it makes 2 buttons not so bad. I don't know why that makes it different but it just seems to to me.

Maybe one button labelled "Status' or "Build Info" which does the same as the ppa button, and one button is just changelog. It's too cluttered though. Too many buttons.

I want to add a do-everything button that does both install-latest and uninstall-old in one shot (and only one pkexec prompt) and I don't want 37 buttons.

It would have to be something most people claim to want to do everytime. If they are always hitting that feature, and it's a lot of users, then it deserves a front-center ui button instead of being one step away from the top surface. If that's the case I'll do that.

**Reply by SamuelMarks** [reply]
Status/Changelog button does sound better than what's currently shown. Seems like a reasonable compromise.

**Comment by StatusCode404** [comment]
Well, for my part, I still prefer having a "Changes" or "Release Notes" button.  But instead of downloading the data.  Just make it a link to go straight to the browser and open the Changes file in browser.

e.g: https://kernel.ubuntu.com/~kernel-ppa/mainline/v6.1.23/CHANGES

Nothing to download.  Just link straight to changes and it opens it on a browser.

**Reply by carlolf** [reply]
`>  I still prefer having a "Changes" or "Release Notes" button.<`
Me too. Why am I forced to open a Browser, then navigate to the CHANGES link instead of just seeing the file my editor? I am not convinced by @bwk77 arguments.

**Reply by bkw777** [reply]
bkw777 is not convinced by carlolf arguments
